### ISSUE TITLE
Implement Builtin Expander (Macro System)

### ISSUE THREAD

**Issue by spy16** [spy16/parens#15]
- [ ] Implement `BuiltinExpander` that expands `Seq` based macro-invocation forms.

Need to figure out how a symbol/value would be identified as a `macro` (Clojure uses a `^:macro` tag)

**Comment by lthibault** [comment]
Perhaps we could use an interface to "tag" values as macros?

```go
type Macro interface{
    Macro()
}
```

We would then wrap a `Seq` in something like this:

```go
type macro struct{ Seq }

func (macro) Macro() {}
```

Then we can just type-assert against the `Macro` type to determine whether macro expansion is required.

**Comment by spy16** [comment]
No, we need to tag the value of the operator to indicate macro or not. We don't need to tag the entire `Seq` form. 

For example, in the invocation form `(my-macro arg1 arg2)`, expander need to be able to figure out that `my-macro` is a macro and then do expansion.

We could introduce a `Macro` type:

```go
type Macro struct {
    Name string    // macro name.
    Args   []string // argument names.
    Body  Any
}
```

In above example,

1. Expander would resolve `my-macro` symbol to get a value of type `Macro`. (If it's not of that type, skip expansion)
2. Create bindings for parameters and the corresponding invocation arguments in the Env. (symbol collision needs to be handled. See [gensym](https://clojuredocs.org/clojure.core/gensym)).
3. Then evaluate the `m.Body` in current `Env`.
4. Return the result as the new form.

**Comment by lthibault** [comment]
Ah gotcha! 

Ok this makes sense. I’ll have a think.

**Comment by lthibault** [comment]
@spy16 I'm going to start working on your proposed implementation.  Just a heads up so that we don't duplicate any work 🙂.

**Comment by lthibault** [comment]
@spy16 Ok, so as much as I hate to admit it, I think I've bitten off more than I can chew!  I think you probably have a much more complete picture of how macro expansion fits into our current architecture, whereas every attempt I've made so far ends up requiring major changes in other parts of the codebase ... a sure sign that I'm Doing It Wrong! 😅 

I'm unassigning myself from this issue, and letting you take over.  My latest attempt can be found in the `feature/macro` branch.  If you have a moment to look, I'd welcome your feedback about what I'm doing incorrectly.

Sorry about this 😞

**Comment by spy16** [comment]
No worries. Let me take a stab at this. 🙂

**Comment by spy16** [comment]
I will try to do this over this weekend. sorry for the slow progress here. have been extremely busy.

**Comment by lthibault** [comment]
No worries, and thanks for the update!  🙂
